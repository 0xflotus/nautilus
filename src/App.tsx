/**
 * ************************************
 *
 * @module  App.tsx
 * @author Joshua Nordstrom
 * @date 3/7/20
 * @description start of the application
 *
 * ************************************
 */

import React, { Component } from 'react';
import { render } from 'react-dom';

// IMPORT REACT CONTAINERS OR COMPONENTS
import Title from './components/Title';
import { Button } from 'react-bootstrap';
import FileSelector from './components/FileUpload';

//bootstrap css
import 'bootstrap/dist/css/bootstrap.min.css';

type AppState = {
  projectName: string;
};

class App extends Component<{}, AppState> {
  state = {
    projectName: 'Nautilus',
  };
  render() {
    return (
      <>
        <Title projectName={this.state.projectName} />
        <Button variant="secondary">hello noobs</Button>
        <FileSelector />
      </>
    );
  }
}

export default App;
