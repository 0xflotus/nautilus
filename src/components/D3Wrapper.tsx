/**
 * ************************************
 *
 * @module  D3Wrapper.tsx
 * @author
 * @date 3/11/20
 * @description Container to hold all the d3 visualation components
 *
 * ************************************
 */
import React from 'react';

import ServicesWrapper from './ServicesWrapper';
import DockerEngine from './DockerEngine';
import HostOS from './HostOS';

type Props = {};

const D3Wrapper: React.FC<Props> = props => {
  return (
    <div className="d3-wrapper">
      <ServicesWrapper />
      <DockerEngine />
      <HostOS />
    </div>
  );
};

export default D3Wrapper;
