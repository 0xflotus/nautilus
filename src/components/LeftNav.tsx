/**
 * ************************************
 *
 * @module  LeftNav.tsx
 * @author
 * @date 3/11/20
 * @description
 *
 * ************************************
 */
import React from 'react';

// IMPORT REACT COMPONENTS
import ServiceInfoDisplay from './ServiceInfoDisplay';
import Title from './Title';
import Button from 'react-bootstrap/Button';
import FileSelector from './FileUpload';

type Props = {
  projectName: string;
};

const LeftNav: React.FC<Props> = ({ projectName }) => {
  return (
    <div className="left-nav">

      <Title projectName={projectName} />
      <FileSelector />
      <ServiceInfoDisplay />
    </div>
  );
};

export default LeftNav;
