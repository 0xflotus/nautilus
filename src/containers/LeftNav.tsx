/**
 * ************************************
 *
 * @module  LeftNav.tsx
 * @author 
 * @date 3/11/20
 * @description 
 * 
 * ************************************
 */
import React from 'react';

type Props = {

}

export const LeftNav: React.FC<Props> = props => {

  return (
    <div>
      
    </div>
  );
}
