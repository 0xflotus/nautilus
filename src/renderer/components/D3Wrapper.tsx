/**
 * ************************************
 *
 * @module  D3Wrapper.tsx
 * @author
 * @date 3/11/20
 * @description Container to hold all the d3 visualation components
 *
 * ************************************
 */
import React from 'react';
import ServicesWrapper from './ServicesWrapper';
import FileSelector from './FileUpload';
import VolumesWrapper from './VolumesWrapper';

//type import
import {
  FileUpload,
  Services,
  SetSelectedContainer,
  Options,
  ReadOnlyObj,
  ViewT,
} from '../App.d';

type Props = {
  fileUpload: FileUpload;
  setSelectedContainer: SetSelectedContainer;
  fileUploaded: boolean;
  services: Services;
  options: Options;
  volumes: Array<ReadOnlyObj>;
  bindMounts: Array<string>;
  view: ViewT;
};

const D3Wrapper: React.FC<Props> = ({
  fileUploaded,
  fileUpload,
  services,
  setSelectedContainer,
  options,
  volumes,
  bindMounts,
  view,
}) => {
  return (
    <div className="d3-wrapper">
      {!fileUploaded ? (
        <FileSelector fileUpload={fileUpload} />
      ) : (
        <>
          <ServicesWrapper
            services={services}
            setSelectedContainer={setSelectedContainer}
            options={options}
            view={view}
          />
          <VolumesWrapper volumes={volumes} bindMounts={bindMounts} />
        </>
      )}
    </div>
  );
};

export default D3Wrapper;
