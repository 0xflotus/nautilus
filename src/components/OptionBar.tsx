/**
 * ************************************
 *
 * @module  OptionBar.tsx
 * @author
 * @date 3/11/20
 * @description Used to display toggle options
 *
 * ************************************
 */
import React, { useState, ReactFragment } from 'react';
import Navbar from 'react-bootstrap/Navbar';
import Nav from 'react-bootstrap/Nav';
import { Options, UpdateOption, UpdateView } from '../App.d';

type Props = {
  view: string;
  options: Options;
  updateView: UpdateView;
  updateOption: UpdateOption;
};

const OptionBar: React.FC<Props> = ({
  view,
  options,
  updateView,
  updateOption,
}) => {
  let dependsViewFlag = false;

  const viewClickToggle = (element: EventTarget & Element): void => {

  }
  const handleViewClick = (e: React.MouseEvent<Element, MouseEvent>) => {
    const element = e.currentTarget
    const v = element.id;

    // Toggle Text Color
    if(element.classList.contains('selected')){
      element.classList.remove('selected')
    } else {
      element.classList.add('selected')
    }

    // Drop Depends_On Option if already in that view
    if (v === 'depends_on') {
      dependsViewFlag = true;
    } else {
      dependsViewFlag = false;
    }

    // Update view in state
    updateView(v);
  };

  const handleOptionClick = (e: React.MouseEvent<Element, MouseEvent>) => {
    const element = e.currentTarget
    const option = e.currentTarget.id;
    console.log('inside handleOptionClick...')
    console.log('options...', options);
    console.log('clicked...', option);

    // Toggle Text Color
    if(element.classList.contains('selected')){
      element.classList.remove('selected')
    } else {
      element.classList.add('selected')
    }

    // Update options in state
    updateOption(option);
  };

  let dependsOption = (
    <Navbar.Text
      className="option"
      id="dependsOnOption"
      onClick={handleOptionClick}
    >
      Depends On
    </Navbar.Text>
  );
  if (view === 'depends_on') {
    dependsOption = (
      <Navbar.Text
        className="option"
        id="dependsOnOption"
        style={{ display: 'none' }}
      >
        Depends On
      </Navbar.Text>
    );
  }

  return (
    <Navbar className="option-bar" variant="dark">
      <Nav className="group-by">
        <Navbar.Text>Group By: </Navbar.Text>
        <Navbar.Text className="option view" id="networks" onClick={handleViewClick}>
          Networks
        </Navbar.Text>
        <Navbar.Text
          className="option view"
          id="depends_on"
          onClick={handleViewClick}
        >
          Depends On
        </Navbar.Text>
      </Nav>
      <Nav className="view-options">
        <Navbar.Text>View Options: </Navbar.Text>
        <Navbar.Text className="option" id="ports" onClick={handleOptionClick}>
          Ports
        </Navbar.Text>
        <Navbar.Text
          className="option"
          id="volumes"
          onClick={handleOptionClick}
        >
          Volumes
        </Navbar.Text>
        {dependsOption}
      </Nav>
    </Navbar>
  );
};

export default OptionBar;
