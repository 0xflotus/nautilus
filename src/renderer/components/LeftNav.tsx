/**
 * ************************************
 *
 * @module  LeftNav.tsx
 * @author
 * @date 3/11/20
 * @description
 *
 * ************************************
 */
import React from 'react';

// IMPORT REACT COMPONENTS
import InfoDropdown from './InfoDropdown';
import FileSelector from './FileUpload';
import Title from './Title';
import { FileUpload, Service } from '../App.d';

type Props = {
  service: Service;
  selectedContainer: string;
  fileUpload: FileUpload;
  fileUploaded: boolean;
};

const LeftNav: React.FC<Props> = ({
  fileUpload,
  fileUploaded,
  selectedContainer,
  service,
}) => {
  return (
    <div className="left-nav">
      <div className="top-half">
        <Title />
        {fileUploaded ? <FileSelector fileUpload={fileUpload} /> : null}
      </div>
      <InfoDropdown selectedContainer={selectedContainer} service={service} />
    </div>
  );
};

export default LeftNav;
