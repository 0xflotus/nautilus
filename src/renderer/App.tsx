/**
 * ************************************
 *
 * @module  App.tsx
 * @author Joshua Nordstrom
 * @date 3/7/20
 * @description start of the application
 *
 * ************************************
 */
//IMPORT LIBRARIES
import React, { Component } from 'react';
import yaml from 'js-yaml';
import { ipcRenderer } from 'electron';

//IMPORT HELPER FUNCTIONS
import convertYamlToState from './helpers/yamlParser';
import { firstTwo } from './helpers/selectAll';
import setGlobalVars from './helpers/setGlobalVars';
import parseUploadError from './helpers/parseUploadError';
import runDockerComposeValidation from '../common/dockerComposeValidation';

// IMPORT STYLES
import './styles/app.scss';

// IMPORT REACT CONTAINERS OR COMPONENTS
import LeftNav from './components/LeftNav';
import OptionBar from './components/OptionBar';
import D3Wrapper from './components/D3Wrapper';

import {
  State,
  FileUpload,
  UpdateOption,
  UpdateView,
  SelectNetwork,
} from './App.d';

const initialState: State = {
  uploadErrors: [],
  selectedContainer: '',
  fileUploaded: false,
  services: {},
  dependsOn: {
    name: 'placeholder',
  },
  networks: {},
  selectedNetwork: '',
  volumes: [],
  volumesClicked: {},
  bindMounts: [],
  bindMountsClicked: {},
  view: 'depends_on',
  options: {
    ports: false,
    volumes: false,
    selectAll: false,
  },
  version: '',
};

class App extends Component<{}, State> {
  constructor(props: {}) {
    super(props);
    this.state = initialState;
  }
  setSelectedContainer = (containerName: string) => {
    this.setState({ ...this.state, selectedContainer: containerName });
  };
  updateView: UpdateView = e => {
    const view = e.currentTarget.id as 'networks' | 'depends_on';
    if (view === 'depends_on') {
      this.setState(state => {
        return {
          ...state,
          view,
          selectedNetwork: '',
        };
      });
    } else {
      this.setState(state => {
        return {
          ...state,
          view,
        };
      });
    }
  };

  updateOption: UpdateOption = e => {
    const option = e.currentTarget.id;
    const selectAllClicked = option === 'selectAll' ? true : false;
    let newState: State = {
      ...this.state,
      options: { ...this.state.options, [option]: !this.state.options[option] },
    };
    if (firstTwo(newState) === true) {
      newState = {
        ...newState,
        options: { ...newState.options, selectAll: true },
      };
    }
    if (firstTwo(this.state) && selectAllClicked) {
      newState = {
        ...this.state,
        options: {
          ports: false,
          volumes: false,
          selectAll: false,
        },
      };
    } else if (!firstTwo(this.state) && selectAllClicked) {
      newState = {
        ...this.state,
        options: {
          ports: true,
          volumes: true,
          selectAll: true,
        },
      };
    } else if (this.state.options.selectAll === true && !firstTwo(newState)) {
      newState = {
        ...newState,
        options: { ...newState.options, selectAll: false },
      };
    }
    this.setState({
      ...newState,
    });
  };

  selectNetwork: SelectNetwork = e => {
    this.setState({ view: 'networks', selectedNetwork: e.currentTarget.value });
  };

  convertAndStoreYamlJSON = (yamlText: string) => {
    const yamlJSON = yaml.safeLoad(yamlText);
    const yamlState = convertYamlToState(yamlJSON);
    setGlobalVars(yamlState.services);
    localStorage.setItem('state', JSON.stringify(yamlState));
    this.setState(Object.assign(initialState, yamlState));
  };

  fileUpload: FileUpload = (file: File) => {
    const fileReader = new FileReader();
    runDockerComposeValidation(file.path).then((validationResults: any) => {
      if (validationResults.error) {
        this.handleFileUploadError(validationResults.error);
      } else {
        fileReader.onload = () => {
          if (fileReader.result) {
            this.convertAndStoreYamlJSON(fileReader.result.toString());
          }
        };
        fileReader.readAsText(file);
      }
    });
  };

  handleFileUploadError = (errorText: Error) => {
    const uploadErrors = parseUploadError(errorText);
    this.setState({
      ...initialState,
      uploadErrors,
      fileUploaded: false,
    });
  };

  componentDidMount() {
    ipcRenderer.on('file-upload-error-within-electron', (event, arg) => {
      this.handleFileUploadError(arg);
    });
    ipcRenderer.on('file-uploaded-within-electron', (event, arg) => {
      this.convertAndStoreYamlJSON(arg);
    });
    const stateJSON = localStorage.getItem('state');
    if (stateJSON) {
      const stateJS = JSON.parse(stateJSON);
      setGlobalVars(stateJS.services);
      this.setState(Object.assign(initialState, stateJS));
    }
  }

  componentWillUnmount() {
    ipcRenderer.removeAllListeners('file-uploaded-within-electron');
    ipcRenderer.removeAllListeners('file-upload-error-within-electron');
  }

  render() {
    return (
      <div className="app-class">
        <div className="draggable"></div>
        <LeftNav
          fileUploaded={this.state.fileUploaded}
          fileUpload={this.fileUpload}
          selectedContainer={this.state.selectedContainer}
          service={this.state.services[this.state.selectedContainer]}
        />
        <div className="main flex">
          <OptionBar
            services={this.state.services}
            view={this.state.view}
            options={this.state.options}
            networks={this.state.networks}
            updateView={this.updateView}
            updateOption={this.updateOption}
            selectNetwork={this.selectNetwork}
            selectedNetwork={this.state.selectedNetwork}
          />
          <D3Wrapper
            uploadErrors={this.state.uploadErrors}
            fileUploaded={this.state.fileUploaded}
            fileUpload={this.fileUpload}
            services={this.state.services}
            setSelectedContainer={this.setSelectedContainer}
            options={this.state.options}
            volumes={this.state.volumes}
            bindMounts={this.state.bindMounts}
            view={this.state.view}
            networks={this.state.networks}
            selectedNetwork={this.state.selectedNetwork}
          />
        </div>
      </div>
    );
  }
}

export default App;
