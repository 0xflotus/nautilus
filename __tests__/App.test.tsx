import React from 'react';
import { shallow, configure, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from '../src/renderer/App';

// IMPORT COMPONENTS
import LeftNav from '../src/renderer/components/LeftNav';
import OptionBar from '../src/renderer/components/OptionBar';
import D3Wrapper from '../src/renderer/components/D3Wrapper';
import { State } from '../src/renderer/App.d';

configure({ adapter: new Adapter() });

describe('Testing App Stateful Component', () => {
  let wrapper: ShallowWrapper<{}, State, App>;
  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  describe('render()', () => {
    it('should render child components', () => {
      expect(wrapper.find(LeftNav)).toHaveLength(1);
      expect(wrapper.find(OptionBar)).toHaveLength(1);
      expect(wrapper.find(D3Wrapper)).toHaveLength(1);
    });
  });

  describe('setSelectedContainer()', () => {
    it('should updated selectedContainer in state', () => {
      wrapper.instance().setSelectedContainer('tester-service');
      expect(wrapper.state().selectedContainer).toBe('tester-service');
      wrapper.instance().setSelectedContainer('api-service');
      expect(wrapper.state().selectedContainer).toBe('api-service');
    });
  });

  describe('updateView()', () => {
    it('should update view in state', () => {
      wrapper.instance().updateView('depends_on');
      expect(wrapper.state().view).toBe('depends_on');
    });
    it('should clear selectedNetwork in state', () => {
      wrapper.state().selectedNetwork = 'dummy-network';
      wrapper.instance().updateView('depends_on');
      expect(wrapper.state().selectedNetwork).toBe('');
    });
  });

  describe('updateOption()', () => {
    beforeEach(() => {
      wrapper.state().options = {
        ports: false,
        volumes: false,
        selectAll: false,
      };
    });
    it('should toggle ports option', () => {
      wrapper.instance().updateOption('ports');
      expect(wrapper.state().options.ports).toBe(true);
      wrapper.instance().updateOption('ports');
      expect(wrapper.state().options.ports).toBe(false);
    });
    it('should toggle volumes option', () => {
      wrapper.instance().updateOption('volumes');
      expect(wrapper.state().options.volumes).toBe(true);
      wrapper.instance().updateOption('volumes');
      expect(wrapper.state().options.volumes).toBe(false);
    });
    it('should toggle selectAll option', () => {
      wrapper.instance().updateOption('selectAll');
      expect(wrapper.state().options.selectAll).toBe(true);
      wrapper.instance().updateOption('selectAll');
      expect(wrapper.state().options.selectAll).toBe(false);
    });
    it('selectAll should toggle ports and options', () => {
      wrapper.instance().updateOption('selectAll');
      expect(wrapper.state().options.ports).toBe(true);
      expect(wrapper.state().options.volumes).toBe(true);
      wrapper.instance().updateOption('selectAll');
      expect(wrapper.state().options.ports).toBe(false);
      expect(wrapper.state().options.volumes).toBe(false);
    });
    it('selectAll should should reflect all being selected or not', () => {
      wrapper.instance().updateOption('ports');
      wrapper.instance().updateOption('volumes');
      expect(wrapper.state().options.selectAll).toBe(true);
      wrapper.instance().updateOption('ports');
      expect(wrapper.state().options.selectAll).toBe(false);
    });
  });

  describe('selectNetwork()', () => {});
});
