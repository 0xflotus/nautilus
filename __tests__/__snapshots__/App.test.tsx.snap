// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App Stateful Component renders correctly 1`] = `
<div
  className="app-class"
>
  <div
    className="draggable"
  />
  <div
    className="left-nav"
  >
    <div
      className="top-half"
    >
      <div
        className="title"
      >
        <img
          src="http://localhost/nautilus_logo.svg"
        />
        <h1>
          Nautilus
        </h1>
      </div>
    </div>
    <div
      className="info-dropdown"
    >
      <h3>
        
      </h3>
      <div
        className="content-wrapper"
      >
        <div
          className="overflow-container"
        >
          <div
            className="overview"
          >
            Please select a container with details to display.
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="main flex"
  >
    <div
      className="option-bar"
    >
      <div
        className="views flex"
      >
        <select
          className="option"
          id="networks"
          name="networks"
          onChange={[Function]}
          value=""
        >
          <option
            disabled={true}
            id="networkHeader"
            value=""
          >
            networks
          </option>
          <option
            className="networkOption"
            id="groupNetworks"
            value="groupNetworks"
          >
            default
          </option>
        </select>
        <span
          className="option selected"
          id="depends_on"
          onClick={[Function]}
        >
          depends on
        </span>
      </div>
      <div
        className="titles flex"
      >
        <h2>
          Views
        </h2>
        <div
          className="vl"
        />
        <h2>
          Options
        </h2>
      </div>
      <div
        className="options flex"
      >
        <span
          className="option"
          id="ports"
          onClick={[Function]}
        >
          ports
        </span>
        <span
          className="option"
          id="volumes"
          onClick={[Function]}
        >
          volumes
        </span>
        <span
          className="option"
          id="selectAll"
          onClick={[Function]}
        >
          select all
        </span>
      </div>
    </div>
    <div
      className="d3-wrapper"
    >
      <div
        className="error-upload-wrapper"
      >
        <div
          className="file-upload"
        >
          <label
            htmlFor="files"
          >
            <div
              className="select-file-button upload-flex"
            >
              <svg
                className="upload-button"
                fill="currentColor"
                height={24}
                size={24}
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 512 512"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M296 384h-80c-13.3 0-24-10.7-24-24V192h-87.7c-17.8 0-26.7-21.5-14.1-34.1L242.3 5.7c7.5-7.5 19.8-7.5 27.3 0l152.2 152.2c12.6 12.6 3.7 34.1-14.1 34.1H320v168c0 13.3-10.7 24-24 24zm216-8v112c0 13.3-10.7 24-24 24H24c-13.3 0-24-10.7-24-24V376c0-13.3 10.7-24 24-24h136v8c0 30.9 25.1 56 56 56h80c30.9 0 56-25.1 56-56v-8h136c13.3 0 24 10.7 24 24zm-124 88c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20zm64 0c0-11-9-20-20-20s-20 9-20 20 9 20 20 20 20-9 20-20z"
                />
              </svg>
              <h5>
                Upload
              </h5>
            </div>
          </label>
          <input
            accept=".yml,.yaml"
            id="files"
            name="yaml"
            onChange={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            type="file"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
